exchanges/gateio/gateio_wrapper.go

```diff
@@ -2392,144 +2183,112 @@ func contractToFundingRate(name string, item asset.Item, fPair currency.Pair, co
+			continue
+		}
+		for _, c := range contracts {
+			if p == currency.EMPTYPAIR { // If not exactly one key provided
+				p, err = g.MatchSymbolWithAvailablePairs(c.contractName(), a, true)
 				if err != nil && !errors.Is(err, currency.ErrPairNotFound) {
-					return nil, err
-				}
-				if !isEnabled {
-					continue
-				}
-				var appendData bool
-				for j := range k {
-					if k[j].Pair().Equal(p) {
-						appendData = true
-						break
-					}
-				}
-				if len(k) > 0 && !appendData {
+					errs = common.AppendError(errs, fmt.Errorf("%w from %s contract %s", err, a, c.contractName()))
 					continue
 				}
-				openInterest := contractResp[i].QuantoMultiplier.Float64() * float64(contractResp[i].PositionSize) * contractResp[i].IndexPrice.Float64()
-				resp = append(resp, futures.OpenInterest{
-					Key: key.ExchangePairAsset{
-						Exchange: g.Name,
-						Base:     p.Base.Item,
-						Quote:    p.Quote.Item,
-						Asset:    a,
-					},
-					OpenInterest: openInterest,
-				})
-			}
-		case asset.Futures:
-			for _, s := range settlementCurrencies {
-				contractResp, err := g.GetAllFutureContracts(ctx, s)
-				if err != nil {
-					return nil, err
-				}
-
-				for i := range contractResp {
-					p, isEnabled, err := g.MatchSymbolCheckEnabled(contractResp[i].Name, a, true)
-					if err != nil && !errors.Is(err, currency.ErrPairNotFound) {
-						return nil, err
-					}
-					if !isEnabled {
+				if len(keys) == 0 { // No keys: All enabled pairs
+					if enabled, err := g.IsPairEnabled(p, a); err != nil {
+						errs = common.AppendError(errs, fmt.Errorf("%w: %s %s", err, a, p))
+						continue
+					} else if !enabled {
 						continue
 					}
-					var appendData bool
-					for j := range k {
-						if k[j].Pair().Equal(p) {
-							appendData = true
-							break
-						}
-					}
-					if len(k) > 0 && !appendData {
+				} else { // More than one key; Any available pair
+					if !slices.ContainsFunc(keys, func(k key.PairAsset) bool { return a == k.Asset && k.Pair().Equal(p) }) {
 						continue
 					}
-					openInterest := contractResp[i].QuantoMultiplier.Float64() * float64(contractResp[i].PositionSize) * contractResp[i].IndexPrice.Float64()
-					resp = append(resp, futures.OpenInterest{
-						Key: key.ExchangePairAsset{
-							Exchange: g.Name,
-							Base:     p.Base.Item,
-							Quote:    p.Quote.Item,
-							Asset:    a,
-						},
-						OpenInterest: openInterest,
-					})
 				}
 			}
+			resp = append(resp, futures.OpenInterest{
+				Key: key.ExchangePairAsset{
+					Exchange: g.Name,
+					Base:     p.Base.Item,
+					Quote:    p.Quote.Item,
+					Asset:    a,
+				},
+				OpenInterest: c.openInterest(),
+			})
 		}
 	}
-	return resp, nil
+	return slices.Clip(resp), errs
+}
+
+type openInterestContract interface {
+	openInterest() float64
+	contractName() string
+}
+
+func (c *FuturesContract) openInterest() float64 {
+	return c.QuantoMultiplier.Float64() * float64(c.PositionSize) * c.IndexPrice.Float64()
+}
```

Comment:
Not you but this looks like notional value not open interest. Open interest can be derived from `GetFutureStats` [ this call ](https://www.gate.io/docs/developers/apiv4/en/#futures-stats)

Reply:
Okay. I think that's outside the scope of this PR, in that it's an existing bug from #1417 which isn't made worse by this change.
It does throw shade on the abstraction work done to support this bug though 😂 

@gloriousCode Could you glance and confirm that you agree this is incorrect?
I'll open the issue and address it in a separate PR, if so.

Reply:
I ran the tests, looked at the numbers, carried the one, compared with the website and confirmed it is wrong